lib/Optimizer/Transforms/ConstPropComplex.cpp

```diff
@@ -0,0 +1,210 @@
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "mlir/Dialect/Complex/IR/Complex.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_CONSTPROPCOMPLEX
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "const-prop-complex"
+
+using namespace mlir;
+
+namespace {
+
+// Replace array ptr casts that throw away the size by a cast to element
+// pointer.
+//
+//%1 = cc.cast %0 : (!cc.ptr<!cc.array<complex<f32> x 4>>) ->
+//! cc.ptr<!cc.array<complex<f32> x ?>>
+// ->
+//%1 = cc.cast %0 : (!cc.ptr<!cc.array<complex<f32> x 4>>) ->
+//! cc.ptr<complex<f32>>
+class CastArrayPtrPattern : public OpRewritePattern<cudaq::cc::CastOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(cudaq::cc::CastOp cast,
+                                PatternRewriter &rewriter) const override {
+
+    auto fromTy = cast.getOperand().getType();
+    auto toTy = cast.getType();
+
+    if (auto ptrFromTy = dyn_cast<cudaq::cc::PointerType>(fromTy)) {
+      if (auto arrayFromTy =
+              dyn_cast<cudaq::cc::ArrayType>(ptrFromTy.getElementType())) {
+        if (auto ptrToTy = dyn_cast<cudaq::cc::PointerType>(toTy)) {
+          if (auto arrayToTy =
+                  dyn_cast<cudaq::cc::ArrayType>(ptrToTy.getElementType())) {
+            if (arrayFromTy.getElementType() == arrayToTy.getElementType()) {
+              auto eleTy = arrayFromTy.getElementType();
+              auto elePtrType = cudaq::cc::PointerType::get(eleTy);
+              rewriter.replaceOpWithNewOp<cudaq::cc::CastOp>(cast, elePtrType,
+                                                             cast.getOperand());
+              return success();
+            }
+          }
+        }
+      }
+    }
+    return failure();
+  }
+};
+
+// Fold complex.create ops if the arguments are constants.
+class ComplexCreatePattern : public OpRewritePattern<complex::CreateOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(complex::CreateOp create,
+                                PatternRewriter &rewriter) const override {
+    auto re = create.getReal();
+    auto im = create.getImaginary();
+    auto reCon = re.getDefiningOp<arith::ConstantOp>();
+    auto imCon = im.getDefiningOp<arith::ConstantOp>();
+    if (reCon && imCon) {
+      auto aa = ArrayAttr::get(
+          rewriter.getContext(),
+          ArrayRef<Attribute>{reCon.getValue(), imCon.getValue()});
+      rewriter.replaceOpWithNewOp<complex::ConstantOp>(create, create.getType(),
+                                                       aa);
+      return success();
+    }
+    return failure();
+  }
+};
+
+// Fold arith.trunc ops if the argument is constant.
```

Comment:
Where are the arith.trunc ops being generated? We need to file an issue against that code. It should be generating cc.cast.

Reply:
I suspect this code: https://github.com/NVIDIA/cuda-quantum/blob/9b9124c9747f1c11fdfd587b4c2bb68596db4832/lib/Frontend/nvqpp/ConvertExpr.cpp#L659

Will file an issue (+ adding complex const prop to the canonicalizer)

Reply:
https://github.com/NVIDIA/cuda-quantum/issues/1955